### ISSUE TITLE
null-coalescing operator ?? not returning the correct value

### ISSUE THREAD

**Issue by pbojanczyk** [dotnet/roslyn#35844]
**Version Used**: 3.0

**Steps to Reproduce**:

1. int? x = default;
2. var y = x ?? default;

**Expected Behavior**: y should be **null**

**Actual Behavior**: y is 0

However if you explicitly specify the cast for default you'll get the correct value like this:
`var y = x ?? (int?)default;`

**Comment by CyrusNajmabadi** [comment]
This looks like the right behavior to me.  `??` takes the type on the left and will unwrap it if it it's a nullable that isn't null (i.e. giving the underlying `int` type if it's non-null).  If it is null, it goes to the right side, but it can see that the intent is to unwrap the left side, so the inferred type on the right would be `int`.  `default(int)` is 0. 

This is sensible since there's no point in returning null on the RHS.  If you wanted to return null when the LHS is null, you would just write it as `var y = x;`.

**Comment by pbojanczyk** [comment]
Thanks for you quick response and explanation.  I still find it a little bit odd that `(int?)default ?? default;` would return 0 instead of null.   Even in Visual Studio when I hover over the left side the Intellisense says it's an int? type and when I hover over the default keyword on the right side, it says it's int? type as well, but when I hover over the ?? operator, it knows that this will return 0

**Comment by CyrusNajmabadi** [comment]
@jcouv for confirmation this behavior is by design.

**Comment by jcouv** [comment]
Tagging @gafter for question while I'm OOF. Otherwise, I'll look next week.

**Comment by gafter** [comment]
The behavior you see is precisely the specified behavior.  See https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#the-null-coalescing-operator bullet three.

Perhaps we should add a new warning when you use `default` on the right-hand-side of a `??` expression?

**Comment by ChaseFlorell** [comment]
> Perhaps we should add a new warning when you use `default` on the right-hand-side of a `??` expression?

yeah, I think a warning would be helpful.

**Comment by gafter** [comment]
I've reopened this as a request for a clear warning about the scenario.

**Comment by svick** [comment]
@pbojanczyk 

>  when I hover over the default keyword on the right side, it says it's int? type

That's not what I see in VS 2019:

![](https://user-images.githubusercontent.com/287848/59067350-a6c49200-88b1-11e9-950b-b8ebe3a7b700.gif)

**Comment by pbojanczyk** [comment]
@svick you're right. It looks like VS 2019 is showing the correct type.  Originally when I reported this issue, I was using VS 2017
